cmd/thanos/store.go

```diff
@@ -104,102 +106,104 @@ func runStore(
+	if err != nil {
+		return errors.Wrap(err, "create bucket client")
+	}
 
-		bkt, err := client.NewBucket(logger, confContentYaml, reg, component)
+	// Ensure we close up everything properly.
+	defer func() {
 		if err != nil {
-			return errors.Wrap(err, "create bucket client")
+			runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
 		}
+	}()
 
-		// Ensure we close up everything properly.
-		defer func() {
-			if err != nil {
-				runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
-			}
-		}()
+	// TODO(bwplotka): Add as a flag?
+	maxItemSizeBytes := indexCacheSizeBytes / 2
 
-		// TODO(bwplotka): Add as a flag?
-		maxItemSizeBytes := indexCacheSizeBytes / 2
+	indexCache, err := storecache.NewIndexCache(logger, reg, storecache.Opts{
+		MaxSizeBytes:     indexCacheSizeBytes,
+		MaxItemSizeBytes: maxItemSizeBytes,
+	})
+	if err != nil {
+		return errors.Wrap(err, "create index cache")
+	}
 
-		indexCache, err := storecache.NewIndexCache(logger, reg, storecache.Opts{
-			MaxSizeBytes:     indexCacheSizeBytes,
-			MaxItemSizeBytes: maxItemSizeBytes,
-		})
-		if err != nil {
-			return errors.Wrap(err, "create index cache")
-		}
+	bs, err := store.NewBucketStore(
+		logger,
+		reg,
+		bkt,
+		dataDir,
+		indexCache,
+		chunkPoolSizeBytes,
+		maxSampleCount,
+		maxConcurrent,
+		verbose,
+		blockSyncConcurrency,
+		filterConf,
+	)
+	if err != nil {
+		return errors.Wrap(err, "create object storage store")
+	}
 
-		bs, err := store.NewBucketStore(
-			logger,
-			reg,
-			bkt,
-			dataDir,
-			indexCache,
-			chunkPoolSizeBytes,
-			maxSampleCount,
-			maxConcurrent,
-			verbose,
-			blockSyncConcurrency,
-			filterConf,
-		)
-		if err != nil {
-			return errors.Wrap(err, "create object storage store")
-		}
+	begin := time.Now()
+	level.Debug(logger).Log("msg", "initializing bucket store")
+	if err := bs.InitialSync(context.Background()); err != nil {
+		return errors.Wrap(err, "bucket store initial sync")
+	}
+	level.Debug(logger).Log("msg", "bucket store ready", "init_duration", time.Since(begin).String())
 
-		begin := time.Now()
-		level.Debug(logger).Log("msg", "initializing bucket store")
-		if err := bs.InitialSync(context.Background()); err != nil {
-			return errors.Wrap(err, "bucket store initial sync")
-		}
-		level.Debug(logger).Log("msg", "bucket store ready", "init_duration", time.Since(begin).String())
-
-		ctx, cancel := context.WithCancel(context.Background())
-		g.Add(func() error {
-			defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
-
-			err := runutil.Repeat(syncInterval, ctx.Done(), func() error {
-				if err := bs.SyncBlocks(ctx); err != nil {
-					level.Warn(logger).Log("msg", "syncing blocks failed", "err", err)
-				}
-				return nil
-			})
-
-			runutil.CloseWithLogOnErr(logger, bs, "bucket store")
-			return err
-		}, func(error) {
-			cancel()
+	ctx, cancel := context.WithCancel(context.Background())
+	g.Add(func() error {
+		defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
```

Comment:
this is a bit confusing, so we were closing `bkt` earlier, why are we closing here again?

Reply:
commented here https://github.com/thanos-io/thanos/pull/1460#discussion_r326884636